File: ast/expression/conversion.go
package expression

import (
	"github.com/panda-io/micro-panda/ast/ast"
	"github.com/panda-io/micro-panda/ast/core"
	"github.com/panda-io/micro-panda/ast/types"
)

type Conversion struct {
	ExpressionBase
	Value ast.Expression
}

func (c *Conversion) Validate(ctx core.Context, expected core.Type) {
	c.Typ = ctx.ResolveType(c.Typ)
	c.Value.Validate(ctx, c.Typ)
	c.Const = c.Value.IsConstant()
	//TO-DO enum convert with int?
	if !((types.IsNumber(c.Typ) && types.IsNumber(c.Value.Type())) || (types.IsPointer(c.Typ) && types.IsPointer(c.Value.Type()))) {
		ctx.Error(c.GetPosition(), "invalid type conversion")
	}
}

File: app/src/main/java/com/github/panda_io/micro_panda/ast/expression/Conversion.java
package com.github.panda_io.micro_panda.ast.expression;

import com.github.panda_io.micro_panda.ast.type.Type;
import com.github.panda_io.micro_panda.ast.Context;

public class Conversion extends Expression {
    public Expression value;

    public void validate(Context context, Type expected) {
        this.type = context.resolveType(this.type);
        this.value.validate(context, null);
        if (this.value.type == null) {
            return;
        }
        this.constant = this.value.isConstant();
        boolean isNumber = this.type.isNumber() && this.value.type.isNumber();
        boolean isPointer = this.type.isPointer() && this.value.type.isPointer();
        if (!(isNumber || isPointer)) {
            context.addError(this.type.getOffset(), "invalid type conversion");
        }
    }
}
